File: 03-day/main.R
library(magrittr)

bits <- readLines('input.txt')

vecToStr <- function(v) v %>% as.character() %>% paste0(collapse = '')

lineToRow <- function(l) {
  vec <- strsplit(l, split = '')[[1]]
  matrix(as.numeric(vec), 1, length(vec))
}

colMode <- function(X) {
  n <- nrow(X)
  n1 <- sum(X[,d])
  (2*n1 >= n)*1
}

X <- sapply(bits, lineToRow) %>% unname() %>% t()

gamma_int <- X %>% colMeans() %>% round()
epsilon_int <- 1 - gamma_int

gamma <- vecToStr(gamma_int)
epsilon <- vecToStr(epsilon_int)

res_a <- strtoi(gamma, base = 2)*strtoi(epsilon, base = 2)

O2 <- X
for (d in 1:ncol(O2)) {
  O2 <- O2[O2[, d] == colMode(O2), , drop = FALSE]
  if (nrow(O2) == 1) break
}

CO2 <- X
for (d in 1:ncol(CO2)) {
  CO2 <- CO2[CO2[, d] == 1 - colMode(CO2), , drop = FALSE]
  if (nrow(CO2) == 1) break
}

O2_dec <- vecToStr(O2) %>% strtoi(base = 2)
CO2_dec <- vecToStr(CO2) %>% strtoi(base = 2)

res_b <- O2_dec*CO2_dec

cat("A ::: Power consumption =", res_a, '\n')
cat("B ::: Life support rating =", res_b, '\n')

File: 03-day/main.py
with open('input.txt', 'r') as f:
    lines = f.readlines()

bits = [line.strip() for line in lines]

w = len(bits[0])
n = len(bits)

gamma = ''
epsilon = ''

for d in range(w):
    digits = [bit[d] for bit in bits]
    if digits.count('0') > digits.count('1'):
        gamma += '0'
        epsilon += '1'
    else:
        gamma += '1'
        epsilon += '0'

res_a = int(gamma, base=2)*int(epsilon, base=2)

print(f"A ::: Power consumption = {res_a}")

# Oxygen generator rating
oxygen_bits = bits.copy()
for d in range(w):
    digits = [bit[d] for bit in oxygen_bits]
    if digits.count('0') > digits.count('1'):
        mode = '0'
    elif digits.count('0') <= digits.count('1'):
        mode = '1'
    oxygen_bits = [bit for bit in oxygen_bits if bit[d] == mode]
    if len(oxygen_bits) == 1:
        break

oxygen_rating = oxygen_bits[0]

# CO2 scrubber rating
co2_bits = bits.copy()
for d in range(w):
    digits = [bit[d] for bit in co2_bits]
    if digits.count('0') > digits.count('1'):
        mode = '1'
    elif digits.count('0') <= digits.count('1'):
        mode = '0'
    co2_bits = [bit for bit in co2_bits if bit[d] == mode]
    if len(co2_bits) == 1:
        break

co2_rating = co2_bits[0]

# 4191984 too low
res_b = int(oxygen_rating, base=2)*int(co2_rating, base=2)
print(f"B ::: Life support rating = {res_b}")
